proposals/0000-visible-forall.rst

```diff
@@ -0,0 +1,1133 @@
+   ``forall`` in ``f``'s type signature. If ``g`` turns out to have a visible
+   dependent type, the argument will be converted to a type; if not, it will be
+   rejected.
+
+2. Punning and a local variable:
+   ::
+
+     f :: forall a. [a] -> [a]
+     f a = g a a
+
+   Here both ``a`` arguments to ``g`` are bound to the inner term-level ``a`` binder (``f``'s
+   argument), regardless of the type of ``g``.
+
+3. Punning and a top-level variable:
+   ::
+
+     a :: Int
+     a = 3
+
+     f :: forall a. [a] -> [a]
+     f x = g a a
+
+   Both ``a`` arguments to ``g`` are bound to term-level bidning for ``a``.  In
+   terms, a term-level binding "wins". If ``g`` turns out to have a visible
+   dependent type, the program will be rejected because ``g``'s first argument
+   is a type, not a term.
+
+4. Punning and types-in-terms:
+   ::
+
+      f :: forall a. [a]->[a]
+      f a = g (a -> (forall b. b -> a)) a
+
+   Again, all those ``a``'s in ``g``'s arguments are bound to the term-level ``a``.
+   The clues that we are in a type, from the ``(->)`` and ``forall``, are not
+   used to change the namespace.
+
+5. Punning and shadowing:
+   ::
+
+     h a = g (forall a. a->a) a
+
+   The ``forall`` binds ``a`` and that binding is seen by the occurrences in ``a->a``.
+   That is, in a term the forall-bound variables are in the data namespace.
+
+6. Built-in syntax:
+   ::
+
+     x1 = g (Int,Bool)
+     x2 = g [Int]
+
+   Here, the built-in syntax occurs in a term-level context, so ``(Int,Bool)``
+   is a promoted pair, and ``[Int]`` is a promoted singleton list.
+
+   One way to change this is to use synonyms from ``GHC.Exts``::
+
+     x1 = g (Tuple2 Int Bool)
+     x2 = g (List Int)
+
+   Another way is to use the ``type`` herald::
+
+     x1 = g (type (Int,Bool))
+     x2 = g (type [Int])
+
+   This is purely a matter of style.
+
+7. The ``@`` changes the meaning of built-in syntax:
+   ::
+
+      a = f @(Int,Bool)
+      b = g  (Int,Bool)
+
+   In ``a``, the argument is the pair type, in ``b`` it is a promoted pair.
+
+   One way to resolve this is to use synonyms from ``GHC.Exts``::
+
+      a = f @(Tuple2 Int Bool)
+      b = g  (Tuple2 Int Bool)
+
+   Another way is to use the ``type`` herald::
+
+      a = f @(Int,Bool)
+      b = g (type (Int,Bool))
+
+8. The ``@`` changes the namespace:
+   ::
+
+     data StrictPair a b = StrictPair !a !b
+
+     x = f  (StrictPair Int Bool)
+     y = g @(StrictPair Int Bool)
+
+   Resolved with the ``type`` herald or by renaming one of the ``StrictPair``
+   constructors.
+
+Effect and Interactions
+-----------------------
+
+* Visible ``forall`` becomes available in types of terms, making them more
+  similar to types of types. There remains a discrepancy that ``forall`` in
+  types of types is actually a relevant quantifier, while the proposed ``forall
+  x ->`` in types of terms is irrelevant. This is to be resolved in the future
+  by making both of them irrelevant.
```

Comment:
relevant?

Reply:
`forall` should probably stay irrelevant, and a new relevant quantifier introduced, e.g. `foreach`. Maybe #370 will come in handy to specify relevance.